ambry-store/src/main/java/com/github/ambry/store/StorageManager.java

```diff
@@ -817,4 +826,165 @@ private void resumeDecommission(String partitionName) throws Exception {
+          resetPartitions(replicasOnFailedDisks);
+          // 5. update disk capacity
+          updateDiskCapacity(healthyDiskCapacity);
+          // 6. Remove disks from the maps.
+          cleanupDisksAndReplicas(newFailedDisks, replicasOnFailedDisks);
+          logger.info("Successfully remove failed disks and replicas from memory when handling disk failure");
+          success = true;
+        } catch (Exception e) {
+          success = false;
+          storeMainMetrics.handleDiskFailureErrorCount.inc();
+        } finally {
+          // 7. exist maintenance mode
+          if (inMaintenanceMode) {
+            if (primaryClusterParticipant.exitMaintenanceMode()) {
+              logger.info("Successfully exit maintenance mode");
+            } else {
+              success = false;
+            }
+          }
+          lock.unlock();
+          lock.close();
+          break;
+        }
+      }
+      if (success) {
+        storeMainMetrics.handleDiskFailureSuccessCount.inc();
+        storeMainMetrics.handleDiskFailureDuration.update(System.currentTimeMillis() - startTime);
+      }
+    }
+
+    private boolean enterMaintenance() {
+      String reason = currentNode.getHostname() + "_" + currentNode.getPort() + "_DISK_FAILURE";
+      if (!primaryClusterParticipant.enterMaintenanceMode(reason)) {
+        throw new IllegalStateException("Failed to enter maintenance mode for reason: " + reason);
+      }
+      return true;
+    }
+
+    private void removeReplicasFromCluster(List<ReplicaId> replicaIds) {
+      if (!primaryClusterParticipant.removeReplicasFromDataNode(replicaIds)) {
+        throw new IllegalStateException(
+            "Failed to remote replicas " + replicaIds + " from cluster when handling disk failure");
+      }
+      logger.info("Partitions are successfully removed from DataNodeConfig of current node when handling disk failure");
+    }
+
+    private void setDiskUnavailable(List<DiskId> diskIds) {
+      diskIds.forEach(diskId -> diskId.setState(HardwareState.UNAVAILABLE));
+      if (!primaryClusterParticipant.setDisksState(diskIds, HardwareState.UNAVAILABLE)) {
+        throw new IllegalStateException("Failed to update disk availability");
+      }
+      logger.info("Successfully update disk availability when handling disk failure");
+    }
+
+    private void resetPartitions(List<ReplicaId> replicasOnFailedDisks) {
+      boolean resetSuccessfully = primaryClusterParticipant.resetPartitionState(replicasOnFailedDisks.stream()
+          .map(ReplicaId::getPartitionId)
+          .map(PartitionId::toPathString)
+          .collect(Collectors.toList()));
+      if (!resetSuccessfully) {
+        throw new IllegalStateException("Failed to reset partitions");
+      }
+      logger.info("Replicas {} are successfully reset when handling disk failure", replicasOnFailedDisks);
+    }
+
+    private void updateDiskCapacity(long healthyDiskCapacity) {
+      final long GB = 1024 * 1024 * 1024;
```

Comment:
Just double check, in Helix, we use 1024 instead of 1000 to calculate the GB etc, right?

Reply:
yes